webhooks: Send different message for force push in GitHub integration.

sbansal1999 opened the pull request:
This PR focuses on sending a different message when a force push occurs on a GitHub branch.

To achieve this, we did the following:

- Added a fixture to simulate the force push event.
- Modified the templates responsible for generating messages for pushes to remove the text `pushed` and include a new field `push_type`, with possible value as either `pushed` or `force-pushed`.
- Checked whether the push was forced and passed the value accordingly.

Fixes: #21969

<hr>

**Screenshots and screen captures:**

| Before| After |
|--------|--------|
|![image](https://user-images.githubusercontent.com/35286603/235319694-3480d5e2-b6cb-4821-83a7-d345958771da.png)| ![image](https://user-images.githubusercontent.com/35286603/235319720-c6a9c542-7509-4539-b595-fb41f9d1cb2a.png)  |

<hr>

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented on zerver/webhooks/github/tests.py:
```diff
@@ -37,6 +37,10 @@ def test_push_delete_branch(self) -> None:
 
+    def test_push_force(self) -> None:
+        expected_message = "sbansal1999 [pushed](https://github.com/sbansal1999/testing-gh/compare/9edbee9dc904...c0dcba5885f8) 2 commits to branch fixing-typo.\n\n* d ([70029144adc](https://github.com/sbansal1999/testing-gh/commit/70029144adc043fbe6f50498ab36acd3c4ba302f))\n* e ([c0dcba5885f](https://github.com/sbansal1999/testing-gh/commit/c0dcba5885f8d9def1f798e3f5eb24e0ace5a030))"
+        self.check_webhook("push__force", "testing-gh / fixing-typo", expected_message)
```
Hmm, this is a force push, but not the typical example of one - say removing the last commit from a branch, where there wouldn't be any new ones. Can you add another fixture for that?

sbansal1999 replied:
You are right, adding a fixture that looks more like a regular force push will be better a idea. Thanks for the suggestion.

timabbott commented on zerver/webhooks/github/view.py:
```diff
@@ -283,6 +283,7 @@ def get_push_commits_body(helper: Helper) -> str:
         get_branch_name_from_ref(payload["ref"].tame(check_string)),
         commits_data,
         deleted=payload["deleted"].tame(check_bool),
+        push_type="force-pushed" if payload["forced"].tame(check_bool) else "pushed",
```
I would spell this "force pushed", and maybe call the parameter `push_type_description` to make clear it's a user-facing string, not a computer-use token like `force-pushed`.

Actually, what I'd actually do is make `get_push_commits_event_message` support an optional `force_push=True|False`, and be responsible for deciding which string to use based on that (only this integration would pass the `True` value at first). And then it'll be much easier to make other Git integrations support this as follow-ups, which you should open an issue for once this is integrated.

sbansal1999 replied:
Changing the parameter to something like a `boolean` is for sure a good idea, as it would also be easy for others to deduce what this parameter will actually do. 

I will make the suggested change and will also make sure to open follow-ups after this one.

sbansal1999 commented:
I have added one more fixture that simulates a more normal force push (removes and adds some commits) and also added a parameter `force_push` to the `get_push_commits_event_message` function as suggested which now tells whether a push is forced or not and then a message is generated accordingly (through the changed templates).

timabbott commented on zerver/lib/webhooks/git.py:
```diff
@@ -89,6 +89,7 @@ def get_push_commits_event_message(
     commits_data: List[Dict[str, Any]],
     is_truncated: bool = False,
     deleted: bool = False,
+    force_push: Optional[bool] = False,
```
Adding an unused parameter as the sole thing a commit doesn't doesn't make sense -- this is not an atomic commit, as the rest of the changes to use it are elsewhere.

sbansal1999 replied:
Rightly said as the changes that use it are somewhere else therefore this commit doesn't makes sense.

I will keep this in mind for the next time.

timabbott commented:
I reworked the commit structure, squashing the fixtures into a single commit (but leaving them separate, so the second functionality change commit could be really clear in showing the functional change via the diff in the tests) and changed `force-pushed` to `force pushed` for the string, that felt more readable.

With those changes, merged, thanks @sbansal1999!

sbansal1999 commented:
Thanks for making those changes and merging them. :heart_decoration: 

I have created #25445 to track the progress of adding the same functionality for other Git Integrations such as GitLab, BitBucket, Azure DevOps, etc.
